### DISCUSSION TITLE
ISO Schematron v/s PURE parity

### DISCUSSION THREAD

**Discussion by hermitokatt** [phax/ph-schematron#148] [General]
I assume (1) that for the same input xml and schematron file validating it with schematron-iso  and schematron-pure delivers the same result (as long as the schematron adheres to the limits of pure). My first question is if this assumption is true? Perhaps the words "different Schematron implementations " in the sentence "shows the different Schematron implementations supported by ph-schematron" already means that then please excuse my lacking grasp of the English language. :)

Assuming (1) is true how is the parity of schematron-iso  and schematron-pure ensured? I had a (quick) look around in the test cases and I could not find anything in regard to this question. I will continue to look around but I would appreciate if you could point me to test cases dealing with this question?

Kind regards,
Stefan

**Comment by phax** [comment]
Hi @Katzor,

Thanks for your questions.
Ad 1) In general, the idea is of course, that they deliver the same results. In practise this is unfortunately not the case. As you can see in issues #88, #91 and #143 the pure implementation struggles with `let` variables that contain XPath expressions. The ISO Schematron and SchXslt approach rely on third-party libraries that convert the SCH to XSLT and apply the XSLT onto the XML to be validated - I hope they both adhere to the same specifcation in the same way, but there is no way that can be guaranteed.

The reason to offer the different implementsions are quite simple:
* ISO Schematron is the reference implementation, but no longer maintained. Additionally it is XSLT based which is in general a bit slower and more memory intensive
* Pure Schematron was basically an approach to overcome the drawbacks of XSLT and stick to Java only - with limited success ;-)
* SchXslt was introduced, because it is currently the only officially maintained Open Source Schematron implementation I am aware of. It is also XSLT based and therefore ships with similar restrictions as the ISO Schematron implementation

There are no explicit test cases that ensure the "cross implementation consistency" - that would be definitively a nice improvement. Feel free to come up with a proposal for a PR - I assume a separate submodule that includes all the existing implementations could be created and make sure the test cases are executed.

**Reply by hermitokatt** [reply]
Hi @phax,

thank you for the explanation! I will try to come up with a proposal for a PR. 👍

**Reply by hermitokatt** [reply]
I started a ticket [here](https://github.com/ecosio/ph-schematron/issues/1).
